presto-main/src/test/java/com/facebook/presto/operator/TestPartitionedOutputOperator.java

```diff
@@ -0,0 +1,340 @@
+        assertTrue(outputSize > simplePage.getSizeInBytes());
+    }
+
+    @Test
+    public void testAddInputForPageWithDictionaryAndReplication()
+    {
+        PartitionedOutputOperator partitionedOutputOperator = createPartitionedOutputOperator(true);
+        Page pageWithDictionary = createPageWithDictionary();
+        partitionedOutputOperator.addInput(pageWithDictionary);
+        OperatorContext operatorContext = partitionedOutputOperator.getOperatorContext();
+        long outputSize = operatorContext.getOutputDataSize().getTotalCount();
+        long outputPositionCount = operatorContext.getOutputPositions().getTotalCount();
+        assertEquals(outputSize, 13_599);
+        assertEquals(outputPositionCount, 1_511);
+        assertTrue(outputSize > pageWithDictionary.getSizeInBytes());
+    }
+
+    @Test
+    public void testAddInputForPageWithRunLengthAndReplication()
+    {
+        PartitionedOutputOperator partitionedOutputOperator = createPartitionedOutputOperator(true);
+        Page pageWithRunLength = createPageWithRunLength();
+        partitionedOutputOperator.addInput(pageWithRunLength);
+        OperatorContext operatorContext = partitionedOutputOperator.getOperatorContext();
+        long outputSize = operatorContext.getOutputDataSize().getTotalCount();
+        long outputPositionCount = operatorContext.getOutputPositions().getTotalCount();
+        assertEquals(outputSize, 13_599);
+        assertEquals(outputPositionCount, 1_511);
+        assertTrue(outputSize > pageWithRunLength.getSizeInBytes());
+    }
+
+    private PartitionedOutputOperator createPartitionedOutputOperator(boolean shouldReplicate)
+    {
+        PartitionFunction partitionFunction = new LocalPartitionGenerator(new InterpretedHashGenerator(ImmutableList.of(BIGINT), new int[] {0}), PARTITION_COUNT);
+        PagesSerdeFactory serdeFactory = new PagesSerdeFactory(new BlockEncodingManager(new TypeRegistry()), false);
+        OutputBuffers buffers = createInitialEmptyOutputBuffers(PARTITIONED);
+        for (int partition = 0; partition < PARTITION_COUNT; partition++) {
+            buffers = buffers.withBuffer(new OutputBuffers.OutputBufferId(partition), partition);
+        }
+        PartitionedOutputBuffer buffer = createPartitionedBuffer(
+                buffers.withNoMoreBufferIds(),
+                new DataSize(Long.MAX_VALUE, BYTE)); // don't let output buffer block
+        buffer.registerLifespanCompletionCallback(ignore -> {});
+        PartitionedOutputOperator.PartitionedOutputFactory operatorFactory = null;
+        if (shouldReplicate) {
+            operatorFactory = new PartitionedOutputOperator.PartitionedOutputFactory(
+                    partitionFunction,
+                    ImmutableList.of(0),
+                    ImmutableList.of(Optional.empty()),
+                    true,
+                    OptionalInt.of(0),
+                    buffer,
+                    new DataSize(1, GIGABYTE));
```

Comment:
Make this DataSize a static final field. Also let's use a smaller size so that flush will happen when we add multiple pages. We can make it 5MB and it will be shared by 512 partitions, so each partition has 10KB memory. We can then add multiple pages to trigger the flushing.

Reply:
Done